Cannot build with clang-16 latest commit

Hello, 
When I try to build the project with clang-16. I got the error 
```ClickHouse/contrib/magic_enum/include/magic_enum.hpp:448:41: error: no matching function for call to 'is_valid'```
And it's very likely related to this [issue](https://github.com/Neargye/magic_enum/issues/204).
More details:
```
-- Cross-compiling for target: 
-- Using compiler:
clang version 16.0.0 (https://github.com/llvm/llvm-project.git e21202dac18ed7f718d26a0e131f96b399b4891c)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/bin
-- Using linker: /usr/bin/ld.lld
-- Using archiver: /usr/local/bin/llvm-ar
-- Using ranlib: /usr/local/bin/llvm-ranlib
-- Using install-name-tool: /usr/local/bin/llvm-install-name-tool
-- Using objcopy: /usr/local/bin/llvm-objcopy
-- Using strip: /usr/local/bin/llvm-strip
-- Using ccache: /usr/bin/ccache (version 3.7.7)
-- HEAD's commit hash f52b6748db0c5076ae3a53390d75dbc71aa88b6a
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
-- CMAKE_BUILD_TYPE is not set, set to default = RelWithDebInfo
-- CMAKE_BUILD_TYPE: RelWithDebInfo
-- Adding .gdb-index via --gdb-index linker option.
-- No official build: A checksum hash will not be added to the clickhouse executable
-- Default libraries: -nodefaultlibs -lgcc  -lc -lm -lrt -lpthread -ldl
-- Some symbols from glibc will be replaced for compatibility
-- Using libunwind: unwind
-- Using exception handler: unwind
-- Unit tests are enabled
-- Building for: Linux x86_64  ;
    USE_STATIC_LIBRARIES=ON
    SPLIT_SHARED_LIBRARIES=OFF
-- Adding contrib module miniselect (configuring with miniselect-cmake)
-- Adding contrib module pdqsort (configuring with pdqsort-cmake)
-- Adding contrib module sparsehash-c11 (configuring with sparsehash-c11-cmake)
-- Adding contrib module abseil-cpp (configuring with abseil-cpp-cmake)
-- Adding contrib module magic_enum (configuring with magic-enum-cmake)
-- Adding contrib module boost (configuring with boost-cmake)
-- Adding contrib module cctz (configuring with cctz-cmake)
-- Packaging with tzdata version: 2022a
-- Adding contrib module consistent-hashing (configuring with consistent-hashing)
-- Adding contrib module dragonbox (configuring with dragonbox-cmake)
-- Adding contrib module vectorscan (configuring with vectorscan-cmake)
-- Adding contrib module jemalloc (configuring with jemalloc-cmake)
-- jemalloc malloc_conf: percpu_arena:percpu,oversize_threshold:0,muzzy_decay_ms:5000,dirty_decay_ms:5000
-- Adding contrib module libcpuid (configuring with libcpuid-cmake)
-- Adding contrib module libdivide (configuring with libdivide)
-- Adding contrib module libmetrohash (configuring with libmetrohash)
-- Adding contrib module lz4 (configuring with lz4-cmake)
-- Adding contrib module murmurhash (configuring with murmurhash)
-- Adding contrib module replxx (configuring with replxx-cmake)
-- Adding contrib module unixodbc (configuring with unixodbc-cmake)
-- Adding contrib module nanodbc (configuring with nanodbc-cmake)
-- Adding contrib module capnproto (configuring with capnproto-cmake)
-- Adding contrib module yaml-cpp (configuring with yaml-cpp-cmake)
-- Adding contrib module re2 (configuring with re2-cmake)
-- Adding contrib module xz (configuring with xz-cmake)
-- Adding contrib module brotli (configuring with brotli-cmake)
-- Adding contrib module double-conversion (configuring with double-conversion-cmake)
-- Adding contrib module boringssl (configuring with boringssl-cmake)
-- Adding contrib module poco (configuring with poco-cmake)
-- Using Poco::Crypto
-- Using Poco::Data::ODBC
-- Adding contrib module croaring (configuring with croaring-cmake)
-- Adding contrib module zstd (configuring with zstd-cmake)
-- ZSTD VERSION 1.5.0
-- Adding contrib module zlib-ng (configuring with zlib-ng-cmake)
-- Adding contrib module bzip2 (configuring with bzip2-cmake)
-- Adding contrib module minizip-ng (configuring with minizip-ng-cmake)
-- Adding contrib module snappy (configuring with snappy-cmake)
-- Adding contrib module rocksdb (configuring with rocksdb-cmake)
-- Adding contrib module thrift (configuring with thrift-cmake)
-- Adding contrib module arrow (configuring with arrow-cmake)
-- Using toolchain file: /home/hanzhou/repos/ClickHouse/cmake/linux/toolchain-x86_64.cmake.
-- CMAKE_CXX_FLAGS:  --gcc-toolchain=/home/hanzhou/repos/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -mavx -fasynchronous-unwind-tables -ffile-prefix-map=/home/hanzhou/repos/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fexperimental-new-pass-manager -w
-- Proceeding with version: 1.12.0.372
-- Adding contrib module avro (configuring with avro-cmake)
-- Adding contrib module protobuf (configuring with protobuf-cmake)
-- Adding contrib module openldap (configuring with openldap-cmake)
-- Adding contrib module grpc (configuring with grpc-cmake)
-- Adding contrib module msgpack-c (configuring with msgpack-c-cmake)
-- Adding contrib module wyhash (configuring with wyhash-cmake)
-- Adding contrib module cityhash102 (configuring with cityhash102)
-- Adding contrib module libfarmhash (configuring with libfarmhash)
-- Adding contrib module icu (configuring with icu-cmake)
-- Adding contrib module h3 (configuring with h3-cmake)
-- Adding contrib module mariadb-connector-c (configuring with mariadb-connector-c-cmake)
-- Adding contrib module googletest (configuring with googletest-cmake)
-- Adding contrib module llvm (configuring with llvm-cmake)
-- LLVM include Directory: /home/hanzhou/repos/ClickHouse/contrib/llvm/llvm/include;/home/hanzhou/repos/ClickHouse/build/contrib/llvm/llvm/include
-- LLVM library Directory: /home/hanzhou/repos/ClickHouse/build/contrib/llvm/llvm
-- LLVM C++ compiler flags: 
-- Native target architecture is X86
-- Threads enabled.
-- Doxygen disabled.
-- Go bindings disabled.
-- Ninja version: 1.10.0
-- Could NOT find OCaml (missing: OCAMLFIND OCAML_VERSION OCAML_STDLIB_PATH) 
-- OCaml bindings disabled.
-- LLVM host triple: x86_64-unknown-linux-gnu
-- LLVM default target triple: x86_64-unknown-linux-gnu
-- Setting native build dir to /home/hanzhou/repos/ClickHouse/build/contrib/llvm/llvm/NATIVE
-- LLVMHello ignored -- Loadable modules not supported on this platform.
-- Targeting X86
-- Targeting AArch64
-- Adding contrib module libxml2 (configuring with libxml2-cmake)
-- Adding contrib module aws;aws-c-common;aws-c-event-stream;aws-checksums (configuring with aws-s3-cmake)
-- Adding contrib module base64 (configuring with base64-cmake)
-- Adding contrib module simdjson (configuring with simdjson-cmake)
-- Adding contrib module rapidjson (configuring with rapidjson-cmake)
-- Adding contrib module fastops (configuring with fastops-cmake)
-- Adding contrib module libuv (configuring with libuv-cmake)
-- Adding contrib module AMQP-CPP (configuring with amqpcpp-cmake)
-- Adding contrib module cassandra (configuring with cassandra-cmake)
-- Adding contrib module curl (configuring with curl-cmake)
-- Adding contrib module azure (configuring with azure-cmake)
By default, if no option is selected, on POSIX, libcurl transport adapter is used.
-- Adding contrib module sentry-native (configuring with sentry-native-cmake)
-- Adding contrib module fmtlib (configuring with fmtlib-cmake)
-- Adding contrib module krb5 (configuring with krb5-cmake)
-- Adding contrib module cyrus-sasl (configuring with cyrus-sasl-cmake)
-- Adding contrib module libgsasl (configuring with libgsasl-cmake)
-- Adding contrib module librdkafka (configuring with librdkafka-cmake)
-- librdkafka with SASL support
-- librdkafka with SSL support
-- Adding contrib module nats-io (configuring with nats-io-cmake)
-- Adding contrib module libhdfs3 (configuring with libhdfs3-cmake)
-- Enable kerberos for HDFS
-- checking compiler: CLANG
-- Checking whether strerror_r returns an int
-- Checking whether strerror_r returns an int -- yes
-- Adding contrib module hive-metastore (configuring with hive-metastore-cmake)
-- Adding contrib module cppkafka (configuring with cppkafka-cmake)
-- Adding contrib module libpqxx (configuring with libpqxx-cmake)
-- Adding contrib module libpq (configuring with libpq-cmake)
-- Adding contrib module NuRaft (configuring with nuraft-cmake)
-- Adding contrib module fast_float (configuring with fast_float-cmake)
-- Adding contrib module datasketches-cpp (configuring with datasketches-cpp-cmake)
-- Adding contrib module hashidsxx (configuring with hashidsxx-cmake)
-- Adding contrib module libstemmer_c (configuring with libstemmer-c-cmake)
-- Adding contrib module wordnet-blast (configuring with wordnet-blast-cmake)
-- Adding contrib module lemmagen-c (configuring with lemmagen-c-cmake)
-- Adding contrib module nlp-data (configuring with nlp-data-cmake)
-- Adding contrib module cld2 (configuring with cld2-cmake)
-- Adding contrib module sqlite-amalgamation (configuring with sqlite-cmake)
-- Adding contrib module s2geometry (configuring with s2geometry-cmake)
-- Adding contrib module c-ares (configuring with c-ares-cmake)
-- Adding contrib module qpl (configuring with qpl-cmake)
-- Not using QPL
-- Performing Test SUPPORTS_CXXFLAG_frame_larger_than=65536
-- Performing Test SUPPORTS_CXXFLAG_frame_larger_than=65536 - Success
-- Performing Test SUPPORTS_CFLAG_frame_larger_than=65536
-- Performing Test SUPPORTS_CFLAG_frame_larger_than=65536 - Success
-- compiler C   = /usr/local/bin/clang  --gcc-toolchain=/home/hanzhou/repos/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -fdiagnostics-color=always -Xclang -fuse-ctor-homing  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -mavx -fasynchronous-unwind-tables -ffile-prefix-map=/home/hanzhou/repos/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries  -fdiagnostics-absolute-paths -fexperimental-new-pass-manager -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -O2 -g -DNDEBUG -O3 -g -gdwarf-4  -fno-pie
-- compiler CXX = /usr/local/bin/clang++  --gcc-toolchain=/home/hanzhou/repos/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -mavx -fasynchronous-unwind-tables -ffile-prefix-map=/home/hanzhou/repos/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fexperimental-new-pass-manager -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-vla -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -O2 -g -DNDEBUG -O3 -g -gdwarf-4  -fno-pie
-- LINKER_FLAGS =  --gcc-toolchain=/home/hanzhou/repos/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 --ld-path=/usr/bin/ld.lld -rdynamic -Wl,--gdb-index -Wl,--build-id=sha1 -no-pie -Wl,-no-pie 
-- /home/hanzhou/repos/ClickHouse/src: Have 27316 megabytes of memory.
        Limiting concurrent linkers jobs to 7 and compiler jobs to 10 (system has 32 logical cores)
-- Will build ClickHouse 22.8.1.1 revision 54465 
-- ClickHouse modes:
-- Server mode: ON
-- Client mode: ON
-- Local mode: ON
-- Self-extracting executable: ON
-- Benchmark mode: ON
-- Extract from config mode: ON
-- Compressor mode: ON
-- Copier mode: ON
-- Format mode: ON
-- Obfuscator mode: ON
-- ODBC bridge mode: ON
-- Library bridge mode: ON
-- ClickHouse install: ON
-- ClickHouse git-import: ON
-- ClickHouse keeper mode: ON
-- ClickHouse keeper-converter mode: ON
-- Clickhouse disks mode: ON
-- ClickHouse su: ON
-- bash_completion will be written to /usr/local/share/bash-completion/completions
-- Target check already exists
-- /home/hanzhou/repos/ClickHouse/utils: Have 27308 megabytes of memory.
        Limiting concurrent linkers jobs to 7 and compiler jobs to OFF (system has 32 logical cores)
-- Configuring done
-- Generating done
```

It's not just about magic_enum: Protobuf also does not compile:
```
/usr/bin/ccache /usr/local/bin/clang++ --target=x86_64-linux-gnu --sysroot=/home/milovidov/work/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc -DGOOGLE_PROTOBUF_CMAKE_BUILD -DHAS_RESERVED_IDENTIFIER -DHAVE_PTHREAD -DHAVE_ZLIB -DSTD_EXCEPTION_HAS_STACK_TRACE=1 -DUNALIGNED_OK -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -D_LIBCPP_ENABLE_THREAD_SAFETY_ANNOTATIONS -I/home/milovidov/work/ClickHouse/build_clang_16/contrib/protobuf -I/home/milovidov/work/ClickHouse/base/glibc-compatibility/memcpy -isystem /home/milovidov/work/ClickHouse/contrib/protobuf/src -isystem /home/milovidov/work/ClickHouse/contrib/libcxx/include -isystem /home/milovidov/work/ClickHouse/contrib/libcxx/src -isystem /home/milovidov/work/ClickHouse/contrib/libcxxabi/include -isystem /home/milovidov/work/ClickHouse/contrib/libunwind/include -isystem /home/milovidov/work/ClickHouse/contrib/zlib-ng -isystem /home/milovidov/work/ClickHouse/build_clang_16/contrib/zlib-ng-cmake --gcc-toolchain=/home/milovidov/work/ClickHouse/cmake/linux/../../contrib/sysroot/linux-x86_64 -std=c++20 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -mavx -fasynchronous-unwind-tables -ffile-prefix-map=/home/milovidov/work/ClickHouse=. -falign-functions=32 -mbranches-within-32B-boundaries -fdiagnostics-absolute-paths -fstrict-vtable-pointers -fexperimental-new-pass-manager -w -O2 -g -DNDEBUG -O3 -g -gdwarf-4  -fno-pie   -D OS_LINUX -Werror -nostdinc++ -std=gnu++20 -MD -MT contrib/protobuf-cmake/CMakeFiles/_libprotobuf.dir/__/protobuf/src/google/protobuf/wire_format_lite.cc.o -MF contrib/protobuf-cmake/CMakeFiles/_libprotobuf.dir/__/protobuf/src/google/protobuf/wire_format_lite.cc.o.d -o contrib/protobuf-cmake/CMakeFiles/_libprotobuf.dir/__/protobuf/src/google/protobuf/wire_format_lite.cc.o -c /home/milovidov/work/ClickHouse/contrib/protobuf/src/google/protobuf/wire_format_lite.cc
/home/milovidov/work/ClickHouse/contrib/protobuf/src/google/protobuf/wire_format_lite.cc:102:9: error: integer value -1 is outside the valid range of values [0, 7] for this enumeration type [-Wenum-constexpr-conversion]
        static_cast<WireFormatLite::WireType>(-1),  // invalid
        ^
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -153,6 +153,8 @@ if (COMPILER_CLANG)
             set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Xclang -fuse-ctor-homing")
         endif()
     endif()
+
+    no_warning(enum-constexpr-conversion) # breaks Protobuf in clang-16
 endif ()
 
 # If compiler has support for -Wreserved-identifier. It is difficult to detect by clang version,
@@ -606,4 +608,3 @@ if (NATIVE_BUILD_TARGETS
         COMMAND ${CMAKE_COMMAND} --build "${NATIVE_BUILD_DIR}" --target ${NATIVE_BUILD_TARGETS}
         COMMAND_ECHO STDOUT)
 endif ()
-

--- a/cmake/warnings.cmake
+++ b/cmake/warnings.cmake
@@ -42,6 +42,7 @@ if (COMPILER_CLANG)
     no_warning(weak-template-vtables)
     no_warning(weak-vtables)
     no_warning(thread-safety-negative) # experimental flag, too many false positives
+    no_warning(enum-constexpr-conversion) # breaks magic-enum library in clang-16
     # TODO Enable conversion, sign-conversion, double-promotion warnings.
 elseif (COMPILER_GCC)
     # Add compiler options only to c++ compiler

